test/unit/connection_string.test.js

```diff
@@ -94,6 +94,13 @@ describe('Connection String', function () {
     expect(options.credentials.mechanism).to.equal(AuthMechanism.MONGODB_GSSAPI);
   });
 
+  it('should provide default authSource when valid AuthMechanism provided', function () {
+    const options = parseOptions(
+      'mongodb+srv://jira-sync.pw0q4.mongodb.net/testDB?authMechanism=MONGODB-AWS&retryWrites=true&w=majority'
+    );
+    expect(options.credentials.source).to.equal('$external');
```

Comment:
Am I understanding correctly that an explicit `authSource` in the TXT record is supposed to be ignored if the auth mechanism is one that always uses a `$external` auth source?

I can see how that makes sense, but it’s also somewhat surprising and doesn’t match my understanding of the SRV spec, because generally options from the TXT record are in between the built-in defaults and the user-specified connection string in terms of priority. (Or the other way around: If we ignore `authSource` from the TXT record for certain auth mechanisms, why is `authSource` not also ignored if it is specified in the connection string?)

Reply:
I took a look at the spec and it does seem that this case isn't explicit spoken on, but the overriding rule is that these mechanisms MUST use $external. 

As for our normal URI parsing, I suppose it should ignore the option, fortunately by not ignoring it this bug does have a workaround by explicitly passing `authSource=$external` the driver will pull it in and the credentials will be approved by the `credentials.validate()`.

Currently our options are not parsed in a particular order, so the source might get set before the mechanism is determined. It might be better if we just coerce to `$external` in the credentials if the mechanism is one of the special ones instead of throwing, maybe? Because of the options parsing strangeness I'm kind of inclined to leave in the "feature" of having the option in the URI override the `$external` rule even if it leads to an error.

I think the real fix is that DNS records should be accurate but it can take time to update, but user's shouldn't have to manually work around that.

Reply:
> I took a look at the spec and it does seem that this case isn't explicit spoken on, but the overriding rule is that these mechanisms MUST use $external.

Then that sounds a bit like this should be clarified in the spec, both for authSource from TXT and from the user-passed connection string? :)

> Because of the options parsing strangeness I'm kind of inclined to leave in the "feature" of having the option in the URI override the `$external` rule even if it leads to an error.

Agreed, it's generally a good idea to error out if conflicting options were explicitly passed as input.

Reply:
DRIVERS-1036 Looks like there's something in the backlog for this, I left a comment describing what we're looking for :)